src/DatumConverter.php

```diff
@@ -0,0 +1,227 @@
+            //          digits are required to recover a double precision number."
+            $previousPrecision = ini_set("precision", 17);
+            $json = json_encode($v);
+            if ($previousPrecision !== false) {
+                ini_set("precision", $previousPrecision);
+            }
+            if ($json === false) {
+                throw new RqlDriverError("Failed to encode document as JSON: " . json_last_error());
+            }
+            return $json;
+        } else {
+            return false;
+        }
+    }
+
+    public function canEncodeAsJson($v)
+    {
+        if (is_null($v) || is_bool($v) || is_int($v) || is_float($v) || is_string($v)) {
+            return true;
+        }
+
+        if (is_array($v)) {
+            foreach ($v as $key => $val) {
+                if (!is_numeric($key) && !is_string($key)) {
+                    return false;
+                }
+                if (!$this->canEncodeAsJson($val)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        return false;
+
+    }
+
+    public function wrapImplicitVar(Query $q)
+    {
+        if ($q->hasUnwrappedImplicitVar()) {
+            return new RFunction(array(new RVar('_')), $q);
+        } else {
+            return $q;
+        }
+    }
+
+    public function nativeToFunction($f)
+    {
+        if (is_object($f) && is_subclass_of($f, '\r\Query')) {
+            return $this->wrapImplicitVar($f);
+        }
+
+        $reflection = new ReflectionFunction($f);
+
+        $args = array();
+        foreach ($reflection->getParameters() as $param) {
+            $args[] = new RVar($param->getName());
+        }
+        $result = $reflection->invokeArgs($args);
+
+        if (!(is_object($result) && is_subclass_of($result, "\\r\\Query"))) {
+            if (!isset($result)) {
+                // In case of null, assume that the user forgot to add a return.
+                // If null is the intended value, r\expr() should be wrapped around the return value.
+                throw new RqlDriverError("The function did not evaluate to a query (missing return?).");
+            } else {
+                $result = $this->nativeToDatum($result);
+            }
+        }
+
+        return new RFunction($args, $result);
+    }
+
+    public function nativeToDatumOrFunction($f)
+    {
+        if (!(is_object($f) && is_subclass_of($f, '\r\Query'))) {
+            try {
+                $f = $this->nativeToDatum($f);
+                if (!is_subclass_of($f, '\r\Datum\Datum')) {
+                    // $f is not a simple datum. Wrap it into a function:
+                    $f = new RFunction(array(new RVar('_')), $f);
+                }
+            } catch (RqlDriverError $e) {
+                $f = $this->nativeToFunction($f);
+            }
+        }
+        return $this->wrapImplicitVar($f);
+    }
+
+    public function systemInfo()
```

Comment:
PHP_RQL_VERSION should also stay in a file of its own if at all possible. It's bad having to modify a proper source code file for each release, just to change the version (among other things it makes the Git history more difficult to interpret).